### PR TITLE
disallow functions in PSyAD (closes #1681)

### LINKED ISSUE #1681: [PSyAD] Test harness incorrect if supplied routine is a Function
All our tests and examples have the target routine as a subroutine. If it is a Function, the test harness still uses `call` and it shouldn't.

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review from @arporter

File: src/psyclone/psyad/tl2ad.py [hunk 1]
```diff
@@ -111,6 +111,15 @@ def generate_adjoint_str(tl_fortran_str, active_variables, create_test=False):
...
     tl_psyir = reader.psyir_from_source(tl_fortran_str)
 
+    # Check there is only one routine (program or subroutine) and that
+    # this routine is not a function,as functions are not supported.
+    routines = tl_psyir.walk(Routine)
+    for routine in routines:
+        if routine.return_symbol:
+            raise NotImplementedError(
+                "PSyAD does not support tangent-linear code written as a "
+                "function. Please re-write as a subroutine.")
```
Comment (arporter) [inline src/psyclone/psyad/tl2ad.py:121]: f"Please re-write '{routine.name}' as a subroutine"

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good.
Tests and examples all OK with compilation.
Docs updated and build fine.
All changes are pylint clean and covered.
Just one tiny request to add the offending function name to the error message and then this can be merged.

File: src/psyclone/psyad/tl2ad.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/psyad/tl2ad.py]: Good idea. Updated.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — APPROVED** [review]
The one requested change has been made.
I'll proceed to merge once I've double-checked compilation tests and examples.

**Pull request merged**